File: node/src/composeParser.ts
import YAML from "yaml";
import type { ImageRef } from "./types.js";

/**
 * Parse a Docker image string into an ImageRef.
 */
export function parseImageRef(imageString: string): ImageRef {
  const raw = imageString;
  let tag: string;
  let namePart: string;

  // Handle digest references (image@sha256:...)
  if (imageString.includes("@")) {
    const atIdx = imageString.indexOf("@");
    namePart = imageString.slice(0, atIdx);
    tag = imageString.slice(atIdx + 1);
  } else {
    const lastSegment = imageString.split("/").pop()!;
    if (lastSegment.includes(":")) {
      const colonPos = imageString.lastIndexOf(":");
      tag = imageString.slice(colonPos + 1);
      namePart = imageString.slice(0, colonPos);
    } else {
      tag = "latest";
      namePart = imageString;
    }
  }

  // Determine registry
  const parts = namePart.split("/");
  let registry: string;
  let remaining: string[];

  if (parts.length >= 2 && (parts[0].includes(".") || parts[0].includes(":"))) {
    registry = parts[0];
    remaining = parts.slice(1);
  } else {
    registry = "docker.io";
    remaining = parts;
  }

  // Determine namespace and name
  let namespace: string;
  let name: string;

  if (remaining.length === 1) {
    namespace = "library";
    name = remaining[0];
  } else if (remaining.length === 2) {
    namespace = remaining[0];
    name = remaining[1];
  } else {
    namespace = remaining[0];
    name = remaining.slice(1).join("/");
  }

  return { registry, namespace, name, tag, raw };
}

/**
 * Parse docker-compose YAML and return list of [serviceName, imageString] pairs.
 */
export function parseCompose(yamlContent: string): [string, string][] {
  const data = YAML.parse(yamlContent);
  const services = data?.services ?? {};
  const results: [string, string][] = [];

  for (const [serviceName, serviceConfig] of Object.entries(services)) {
    if (
      serviceConfig !== null &&
      typeof serviceConfig === "object" &&
      "image" in (serviceConfig as Record<string, unknown>)
    ) {
      results.push([serviceName, (serviceConfig as Record<string, unknown>).image as string]);
    }
  }

  return results;
}

File: python/code_provenance/compose_parser.py
import yaml
from code_provenance.models import ImageRef


def parse_image_ref(image_string: str) -> ImageRef:
    """Parse a Docker image string into an ImageRef."""
    raw = image_string
    digest = None

    # Handle digest references (image@sha256:... or image:tag@sha256:...)
    if "@" in image_string:
        name_part, digest = image_string.split("@", 1)
        # Check if there's a tag before the digest (image:tag@sha256:...)
        last_segment = name_part.split("/")[-1]
        if ":" in last_segment:
            colon_pos = name_part.rfind(":")
            tag = name_part[colon_pos + 1:]
            image_string = name_part[:colon_pos]
        else:
            tag = digest
            image_string = name_part
    elif ":" in image_string.split("/")[-1]:
        colon_pos = image_string.rfind(":")
        tag = image_string[colon_pos + 1:]
        image_string = image_string[:colon_pos]
    else:
        tag = "latest"

    # Determine registry
    parts = image_string.split("/")
    if len(parts) >= 2 and ("." in parts[0] or ":" in parts[0]):
        registry = parts[0]
        remaining = parts[1:]
    else:
        registry = "docker.io"
        remaining = parts

    # Determine namespace and name
    if len(remaining) == 1:
        namespace = "library"
        name = remaining[0]
    elif len(remaining) == 2:
        namespace = remaining[0]
        name = remaining[1]
    else:
        namespace = remaining[0]
        name = "/".join(remaining[1:])

    return ImageRef(
        registry=registry,
        namespace=namespace,
        name=name,
        tag=tag,
        digest=digest,
        raw=raw,
    )


def parse_compose(yaml_content: str) -> list[tuple[str, str]]:
    """Parse docker-compose YAML and return list of (service_name, image_string)."""
    data = yaml.safe_load(yaml_content)
    if not isinstance(data, dict):
        return []
    services = data.get("services", {}) or {}
    results = []
    for service_name, service_config in services.items():
        if isinstance(service_config, dict) and "image" in service_config:
            results.append((service_name, service_config["image"]))
    return results
